bundles/org.openhab.core.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/internal/ClientCallback.java

```diff
@@ -64,14 +66,32 @@ public synchronized void connectionLost(@Nullable Throwable exception) {
         }
     }
 
-    @Override
-    public void deliveryComplete(IMqttDeliveryToken token) {
-        logger.trace("Message with id {} delivered.", token.getMessageId());
+    public void messageArrived(Mqtt3Publish message) {
+        byte[] payload = message.getPayloadAsBytes();
+        String topic = message.getTopic().toString();
+        logger.trace("Received message on topic '{}' : {}", topic, new String(payload));
+        List<MqttMessageSubscriber> matches = new ArrayList<>();
+        synchronized (subscribers) {
+            subscribers.values().forEach(subscriberList -> {
+                if (subscriberList.topicMatch(topic)) {
+                    logger.trace("Topic match for '{}' using regex {}", topic, subscriberList.getTopicRegexPattern());
+                    subscriberList.forEach(consumer -> matches.add(consumer));
+                } else {
+                    logger.trace("No topic match for '{}' using regex {}", topic,
+                            subscriberList.getTopicRegexPattern());
+                }
+            });
+        }
+        try {
+            matches.forEach(subscriber -> subscriber.processMessage(topic, payload));
+        } catch (Exception e) {
+            logger.error("MQTT message received. MqttMessageSubscriber#processMessage() implementation failure", e);
+        }
     }
 
-    @Override
-    public void messageArrived(String topic, MqttMessage message) {
-        byte[] payload = message.getPayload();
+    public void messageArrived(Mqtt5Publish message) {
+        byte[] payload = message.getPayloadAsBytes();
+        String topic = message.getTopic().toString();
```

Comment:
Right now there is code duplication for those two `messageArrived` methods. How about an internal third one `void messageArrived(byte[] payload, String topic)` that gets called by the two others?

Reply:
Done